File: Sample/sample_cs/ObjectSystem/Object2DComponent.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace sample_cs.ObjectSystem
{
	/// <summary>
	/// コンポーネントを用いてオブジェクトを変化させるサンプル。
	/// </summary>
	class Object2DComponent : ISample
	{
		class RotateComponent : asd.Object2DComponent
		{
			protected override void OnUpdate()
			{
				// 毎フレーム、オブジェクトの角度を回転させる
				Owner.Angle += 2;
			}
		}

		public void Run()
		{
			// aceの初期化
			asd.Engine.Initialize("Object2DComponent", 640, 480, new asd.EngineOption());

			// 使用するシーン、レイヤー、オブジェクト、コンポーネントを生成
			var scene = new asd.Scene();
			var layer = new asd.Layer2D();
			var obj = new asd.TextureObject2D();
			var component = new RotateComponent();


			// シーンを変更
			asd.Engine.ChangeScene(scene);

			// シーンにレイヤーを追加
			scene.AddLayer(layer);

			// レイヤーにオブジェクトを追加
			layer.AddObject(obj);

			// オブジェクトに回転コンポーネントをコンポーネント名"Rotate"で追加
			obj.AddComponent(component, "Rotate");


			// テクスチャをロード
			var texture = asd.Engine.Graphics.CreateTexture2D("Data/Texture/Block1.png");

			// オブジェクトに画像をセット
			obj.Texture = texture;

			// オブジェクトの位置を指定
			obj.Position = new asd.Vector2DF(320, 240);

			while(asd.Engine.DoEvents())
			{
				asd.Engine.Update();
			}

			asd.Engine.Terminate();
		}
	}
}

File: Sample/sample_cpp/ObjectSystem/Object2DComponent.cpp
#include <Base.h>

class RotatingComponent : public asd::Object2DComponent
{
public:
	void OnUpdate()
	{
		// 毎フレーム、オブジェクトの角度を回転させる
		GetOwner()->SetAngle(GetOwner()->GetAngle() + 2);
	}
};

void Object2DComponent()
{
	// aceを初期化する
	asd::Engine::Initialize(asd::ToAString("Object2DComponent").c_str(), 640, 480, asd::EngineOption());

	// 使用するシーン、レイヤー、オブジェクト、コンポーネントを生成
	auto scene = std::make_shared<asd::Scene>();
	auto layer = std::make_shared<asd::Layer2D>();
	auto object = std::make_shared<asd::TextureObject2D>();
	auto component = std::make_shared<RotatingComponent>();


	// シーンを変更
	asd::Engine::ChangeScene(scene);

	// シーンにレイヤーを追加
	scene->AddLayer(layer);

	// レイヤーにオブジェクトを追加
	layer->AddObject(object);

	// オブジェクトに回転コンポーネントをコンポーネント名"Rotate"で追加
	object->AddComponent(component, asd::ToAString("Rotate"));


	// グラフィックスオブジェクトを取得
	auto graphics = asd::Engine::GetGraphics();

	// テクスチャをロード
	auto texture = graphics->CreateTexture2D(asd::ToAString("Data/Texture/Block1.png").c_str());

	// オブジェクトに画像をセット
	object->SetTexture(texture);

	// オブジェクトの位置を指定
	object->SetPosition(asd::Vector2DF(320, 240));


	while (asd::Engine::DoEvents())
	{
		asd::Engine::Update();
	}

	asd::Engine::Terminate();
}